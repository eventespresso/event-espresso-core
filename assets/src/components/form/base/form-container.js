/**
 * Internal imports
 */
import './form-container.css';

/**
 * External imports
 */
import classNames from 'classnames';


/**
 * FormContainer is a wrapper that hides the form while it is loading,
 * then fades it in once loading is complete
 *
 * @function
 * @param {Object} form
 * @param {string} htmlClass
 * @param {boolean} loading
 * @return {Object} form
 */
export const FormContainer = ( { children, htmlClass, loading } ) => {
	const containerClass = classNames( {
		'ee-form-container-div': true,
		[ `${ htmlClass }` ]: htmlClass,
		'ee-form-fade-in': ! loading,
	} );
	return (
		<div className={ containerClass } >
			{ ! loading && children }
		</div>
	);
};
