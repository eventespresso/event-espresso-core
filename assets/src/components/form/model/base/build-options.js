/**
 * External dependencies
 */
import {
	isEmpty,
	isFunction,
	isUndefined,
	reduce,
	isArray,
	isMap,
} from 'lodash';
import warning from 'warning';

/**
 * Receives an array|Object|Map of event entities and returns an array of simple
 * objects that can be passed along to the options array used for a
 * SelectControl component
 *
 * @param { Map|Object|Array }  entities
 * @param { Object } optionsEntityMap Object of key value pairs where values
 * are the 'label:value' pairings used to populate the Select input.
 * @param { string } mapSelection Determines which optionEntityMap pairing to
 * use.
 * @return { Array } Returns an array of simple objects formatted for any
 * select control that receives its options in the format of an array of objects
 * with label and value keys.
 */
const buildOptions = (entities, optionsEntityMap, mapSelection = 'default') => {
	if (isEmpty(entities)) {
		return [];
	}
	entities = isMap(entities) ? Array.from(entities.values()) : entities;
	if (isEmpty(optionsEntityMap)) {
		warning(
			false,
			'A valid optionsEntityMap must be supplied ' +
				'in order to build options for a ModelSelect component',
			optionsEntityMap
		);
		return [];
	}
	if (isUndefined(optionsEntityMap[mapSelection])) {
		warning(
			false,
			'A valid optionsEntityMap[ mapSelection ] must be supplied ' +
				'in order to build options for a ModelSelect component',
			optionsEntityMap,
			mapSelection
		);
		return [];
	}
	// if requested mapSelection exists then use that
	let map = optionsEntityMap[mapSelection];

	// if map is function then simply pass through entities to that function
	if (isFunction(map)) {
		const options = map(entities);
		if (isArray(options)) {
			return options;
		}
		// set map to empty object because the function on it returned something
		// other than an array
		map = {};
		warning(
			false,
			'The optionsEntityMap for ' +
				mapSelection +
				'was a function but' +
				' it did not return an array of options.  Make sure the function' +
				' returns an array of simple objects with label and value keys.'
		);
	}
	if (isEmpty(map)) {
		return [];
	}
	const reducerCallback = (options, entity) => {
		const label = isFunction(map.label)
			? map.label(entity)
			: entity[map.label];
		const value = isFunction(map.value)
			? map.value(entity)
			: entity[map.value];
		if (label && value) {
			options.push({ label, value });
		}
		return options;
	};
	return !isArray(entities)
		? entities.reduce(reducerCallback, [])
		: reduce(entities, reducerCallback, []);
};

export default buildOptions;
