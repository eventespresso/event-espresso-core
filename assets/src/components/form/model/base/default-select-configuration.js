/**
 * External imports
 */
import PropTypes from 'prop-types';
import { __ } from '@eventespresso/i18n';

export const REACT_SELECT_TYPES = {
	'aria-describedby': PropTypes.string,
	'aria-label': PropTypes.string,
	'aria-labelledby': PropTypes.string,
	autoFocus: PropTypes.bool,
	backspaceRemovesValue: PropTypes.bool,
	blurInputOnSelect: PropTypes.bool,
	captureMenuScroll: PropTypes.bool,
	className: PropTypes.string,
	classNamePrefix: PropTypes.string,
	closeMenuOnSelect: PropTypes.bool,
	components: PropTypes.object,
	controlShouldRenderValue: PropTypes.bool,
	delimiter: PropTypes.string,
	escapeClearsValue: PropTypes.bool,
	filterOption: PropTypes.func,
	formatGroupLabel: PropTypes.func,
	formatOptionLabel: PropTypes.func,
	getOptionLabel: PropTypes.func,
	getOptionValue: PropTypes.func,
	hideSelectedOptions: PropTypes.bool,
	id: PropTypes.string,
	inputValue: PropTypes.string,
	inputId: PropTypes.string,
	instanceId: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
	isClearable: PropTypes.bool,
	isDisabled: PropTypes.bool,
	isLoading: PropTypes.bool,
	isOptionDisabled: PropTypes.func,
	isOptionSelected: PropTypes.func,
	isMulti: PropTypes.bool,
	isSearchable: PropTypes.bool,
	loadingMessage: PropTypes.func,
	minMenuHeight: PropTypes.number,
	maxMenuHeight: PropTypes.number,
	menuIsOpen: PropTypes.bool,
	menuPlacement: PropTypes.oneOf(['auto', 'bottom', 'top']),
	menuPosition: PropTypes.oneOf(['absolute', 'fixed']),
	menuPortalTarget: PropTypes.element,
	menuShouldBlockScroll: PropTypes.bool,
	menuShouldScrollIntoView: PropTypes.bool,
	name: PropTypes.string,
	noOptionsMessage: PropTypes.func,
	onBlur: PropTypes.func,
	onChange: PropTypes.func,
	onFocus: PropTypes.func,
	onInputChange: PropTypes.func,
	onKeyDown: PropTypes.func,
	onMenuOpen: PropTypes.func,
	onMenuClose: PropTypes.func,
	onMenuScrollToTop: PropTypes.func,
	onMenuScrollToBottom: PropTypes.func,
	openMenuOnFocus: PropTypes.bool,
	openMenuOnClick: PropTypes.bool,
	options: PropTypes.array,
	pageSize: PropTypes.number,
	placeholder: PropTypes.string,
	screenReaderStatus: PropTypes.func,
	styles: PropTypes.shape({
		clearIndicator: PropTypes.func,
		container: PropTypes.func,
		control: PropTypes.func,
		dropdownIndicator: PropTypes.func,
		group: PropTypes.func,
		groupHeading: PropTypes.func,
		indicatorsContainer: PropTypes.func,
		indicatorSeparator: PropTypes.func,
		input: PropTypes.func,
		loadingIndicator: PropTypes.func,
		loadingMessageCSS: PropTypes.func,
		menu: PropTypes.func,
		menuList: PropTypes.func,
		menuPortal: PropTypes.func,
		multiValue: PropTypes.func,
		multiValueLabel: PropTypes.func,
		multiValueRemove: PropTypes.func,
		noOptionsMessageCSS: PropTypes.func,
		option: PropTypes.func,
		placeholder: PropTypes.func,
		singleValue: PropTypes.func,
		valueContainer: PropTypes.func,
	}),
	tabIndex: PropTypes.string,
	tabSelectsValue: PropTypes.bool,
	value: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
};

export const REACT_SELECT_DEFAULTS = {
	isClearable: true,
	isLoading: true,
	placeholder: __('Selectâ€¦', 'event_espresso'),
};
