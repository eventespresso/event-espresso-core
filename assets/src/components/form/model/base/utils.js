/**
 * External imports
 */
import { upperFirst, camelCase } from 'lodash';

/**
 * Return a name for a model select component.
 *
 * @param {string} modelName
 * @return {string} Receives something like 'message_template' and returns
 * MessageTemplateSelect
 */
export const modelSelectName = (modelName) => {
	return upperFirst(camelCase(modelName)) + 'Select';
};
