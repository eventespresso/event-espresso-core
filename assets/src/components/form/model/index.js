export {
	ModelSelect,
	default as ModelEnhancedSelect,
} from './base/model-select';
export * from './select';
export * from './base/build-options';
export { default as createModelSelect } from './base/create-model-select';
