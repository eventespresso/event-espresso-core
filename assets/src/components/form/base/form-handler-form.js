/**
 * External imports
 */
import { isEmpty } from 'lodash';
import { useCallback, useMemo } from '@wordpress/element';

/**
 * Internal imports
 */
import { FormContainer } from './form-container';
import { FormDataDebugDump } from './form-data-debug-dump';
import FormPlaceholder from './form-placeholder';
import FormSubmitButton from './form-submit-button';
import FormCancelButton from './form-cancel-button';

const nullFunc = () => null;

const FormHandlerForm = ( {
	FormComponent,
	form,
	initialValues,
	currentValues,
	handleSubmit,
	submitting,
	pristine,
	invalid,
	loading,
	loadingNotice,
	showSubmit,
	submitButtonText,
	showCancel,
	cancelButtonText,
	formChanges = false,
	resetHandler = nullFunc,
	...formProps
} ) => {
	pristine = pristine && ! formChanges;
	currentValues = isEmpty( currentValues ) ? initialValues : currentValues;
	const formReset = useCallback(
		( event ) => {
			if ( resetHandler !== nullFunc ) {
				resetHandler( event );
			}
			form.reset( event );
		},
		[ resetHandler, form.reset ]
	);
	const submitButton = useMemo(
		() => showSubmit ?
			<FormSubmitButton
				buttonText={ submitButtonText }
				submitting={ submitting }
				disabled={ pristine || invalid }
			/> :
			null,
		[ showSubmit, submitButtonText, submitting, pristine, invalid ]
	);
	const cancelButton = useMemo(
		() => showCancel ?
			<FormCancelButton
				buttonText={ cancelButtonText }
				onClick={ formReset }
				pristine={ pristine }
				submitting={ submitting }
			/> :
			null,
		[ showCancel, cancelButtonText, formReset, submitting, pristine ]
	);
	return (
		<form onSubmit={ handleSubmit }>
			<FormPlaceholder
				loading={ loading }
				notice={ loadingNotice }
			/>
			<FormContainer loading={ loading }>
				<FormComponent
					updateField={ form.change }
					touchField={ form.blur }
					initialValues={ initialValues }
					currentValues={ currentValues }
					submitButton={ submitButton }
					cancelButton={ cancelButton }
					formReset={ formReset }
					pristine={ pristine }
					{ ...formProps }
				/>
			</FormContainer>
			<FormDataDebugDump values={ currentValues } />
		</form>
	);
};

export default FormHandlerForm;
