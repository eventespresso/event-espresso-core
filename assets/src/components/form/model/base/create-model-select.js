/**
 * External imports
 */
import { Component } from '@wordpress/element';

/**
 * Internal imports
 */
import ModelSelect from './model-select';
import { modelSelectName } from './utils';
import PropTypes from 'prop-types';

const defaultPropTypes = {
	selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
	onSelect: PropTypes.func,
	label: PropTypes.string,
};

/**
 * Creates a Model Select component using the given arguments.
 *
 * @param {string} modelName
 * @param {Object} defaultProps
 * @param {Object} propTypes
 * @return {ModelSelectComponent} A model select component.
 */
const createModelSelect = (modelName, defaultProps = {}, propTypes = {}) => {
	class ModelSelectComponent extends Component {
		/**
		 * @type {string}
		 */
		static modelName = modelName;

		/**
		 * @type {Object}
		 */
		static defaultProps = defaultProps;

		/**
		 * @type {Object}
		 */
		static propTypes = {
			...defaultPropTypes,
			...propTypes,
		};

		render() {
			const { selected, onSelect } = this.props;
			let { queryData } = this.props;
			const selectOptions = {
				selectConfiguration: {
					defaultValue: selected,
					onChange: onSelect,
					...this.props.selectConfiguration,
				},
			};

			if (defaultProps.queryData && queryData) {
				queryData = {
					...defaultProps.queryData,
					...queryData,
				};
			}

			const props = {
				...this.props,
				...selectOptions,
				queryData,
				modelName: this.constructor.modelName,
			};
			return <ModelSelect {...props} />;
		}
	}
	ModelSelectComponent.displayName = modelSelectName(modelName);
	return ModelSelectComponent;
};

export default createModelSelect;
