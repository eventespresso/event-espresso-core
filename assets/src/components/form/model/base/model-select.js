/**
 * External imports
 */
import Select from 'react-select';
import { Component, Fragment } from '@wordpress/element';
import {
	isEmpty,
	uniqueId,
	find,
	isUndefined,
	isFunction,
	isMap,
} from 'lodash';
import PropTypes from 'prop-types';
import isShallowEqual from '@wordpress/is-shallow-equal';

/**
 * WP dependencies
 */
import { withSelect } from '@wordpress/data';

/**
 * Internal imports
 */
import buildOptions from './build-options';
import { MODEL_NAMES } from '../../../../data/model/index';
import {
	REACT_SELECT_DEFAULTS,
	REACT_SELECT_TYPES,
} from './default-select-configuration';
import {
	ALLOWED_ORDER_VALUES,
	QUERY_ORDER_DESC,
} from '../../../../data/model/base';

/**
 * ModelSelect component.
 * This is a component that will generate a react-select input for a given
 * model and its entities (provided via props).
 *
 * @see https://deploy-preview-2289--react-select.netlify.com/props#prop-types
 *   for options that can be passed through via the selectConfiguration prop.
 *
 * @param { Object } selectConfiguration 	An object containing options for the
 *                                          react-select component.
 * @param { Array } modelEntities          	Array of model entities
 * @param { string } modelName              The name of the Model the entities
 *                                          belong to.
 * @param {Function} mapOptionsCallback  	This function will receive by default
 * 											the modelEntities and optionsEntityMap,
 * 											and is expected to return an array of options
 * 											to be used for the react-select component.
 * @param { Object } optionsEntityMap    	Object of key value pairs where values are the
 *                                        	'label:value' pairings used by `mapOptionsCallback`
 * @param { string } mapSelection        	Determines which optionEntityMap pairing to use
 */
export class ModelSelect extends Component {
	state = {};

	static propTypes = {
		selectConfiguration: PropTypes.shape({
			...REACT_SELECT_TYPES,
		}),
		modelEntities: PropTypes.array,
		modelName: PropTypes.oneOf(MODEL_NAMES).isRequired,
		mapOptionsCallback: PropTypes.func,
		optionsEntityMap: PropTypes.object,
		mapSelection: PropTypes.string,
		queryData: PropTypes.shape({
			limit: PropTypes.number,
			orderBy: PropTypes.string,
			order: PropTypes.oneOf(ALLOWED_ORDER_VALUES),
		}),
		getQueryString: PropTypes.func,
		label: PropTypes.string,
	};

	static defaultProps = {
		selectConfiguration: {
			...REACT_SELECT_DEFAULTS,
			name: uniqueId('model-select-'),
		},
		modelEntities: [],
		mapOptionsCallback: buildOptions,
		mapSelection: 'default',
		queryData: {
			limit: 100,
			order: QUERY_ORDER_DESC,
		},
		label: '',
		getQueryString: () => '',
	};

	constructor(props) {
		super(props);
		this.state = ModelSelect.setStateFromProps(props);
	}

	/**
	 * Sets the state from provided props
	 *
	 * @param {Object} props
	 * @return {{
	 * isClearable,
	 * isLoading,
	 * placeholder,
	 * options: (*|Array),
	 * value: *
	 * }}
	 * Object to replace state.
	 */
	static setStateFromProps(props) {
		const { selectConfiguration } = props;
		const options = ModelSelect.getOptions(props);
		const selectedValue = ModelSelect.getOptionObjectForValue(
			selectConfiguration.defaultValue,
			options
		);
		const updated = {
			options,
			value: selectedValue,
		};

		return {
			...REACT_SELECT_DEFAULTS,
			...selectConfiguration,
			...updated,
		};
	}

	/**
	 * Sets up options for the select control from the incoming props.
	 *
	 * @param {Object} props
	 * @return {Array<Object>} Array of options for select control
	 */
	static getOptions(props) {
		const {
			mapOptionsCallback,
			modelEntities,
			optionsEntityMap,
			mapSelection,
		} = props;
		if (isFunction(mapOptionsCallback) && !isEmpty(modelEntities)) {
			return mapOptionsCallback(
				modelEntities,
				optionsEntityMap,
				mapSelection
			);
		}
		return [];
	}

	/**
	 * Given a value, returns the corresponding option object.
	 *
	 * @param {*} value
	 * @param {Array<Object>} options
	 * @return {Object|null} The option object for the given value or null.
	 */
	static getOptionObjectForValue(value, options) {
		if (!isEmpty(options)) {
			const match = find(options, function (option) {
				return option.value === value;
			});
			return !isUndefined(match) ? match : null;
		}
		return null;
	}

	/**
	 * Helper method for determining whether state should be updated.
	 * Criteria is whether the keys from the incoming entities are equivalent
	 * (equivalency means in same order as well)
	 *
	 * @param {Object} prevProps
	 * @param {Object} nextProps
	 * @return {boolean} True means state should be updated.
	 */
	static shouldUpdateState(prevProps, nextProps) {
		if (prevProps === nextProps) {
			return false;
		}

		const { selectConfiguration: prevConfiguration } = prevProps;
		const { selectConfiguration: nextConfiguration } = nextProps;

		// if defaultValue has changed (selected value) then update state
		if (prevConfiguration.defaultValue !== nextConfiguration.defaultValue) {
			return true;
		}

		if (prevConfiguration.isLoading !== nextConfiguration.isLoading) {
			return true;
		}

		// shallow compare of keys but only if we can
		if (
			!isMap(prevProps.modelEntities) ||
			!isMap(nextProps.modelEntities)
		) {
			return true;
		}

		if (
			!isShallowEqual(
				Array.from(prevProps.modelEntities.keys()),
				Array.from(nextProps.modelEntities.keys())
			)
		) {
			return true;
		}
	}

	componentDidMount() {
		this.setState(ModelSelect.setStateFromProps(this.props));
	}

	componentDidUpdate(prevProps) {
		if (ModelSelect.shouldUpdateState(prevProps, this.props)) {
			this.setState(ModelSelect.setStateFromProps(this.props));
		}
	}

	/**
	 * Returns the label for the select control
	 *
	 * @return {string} The label to use.
	 */
	getSelectLabel() {
		const { label, selectConfiguration } = this.props;
		return label ? (
			<label htmlFor={selectConfiguration.name}>{label}</label>
		) : (
			''
		);
	}

	render() {
		return (
			<Fragment>
				{this.getSelectLabel()}
				<Select {...this.state} />
			</Fragment>
		);
	}
}

/**
 * The ModelSelect Component wrapped in the `withSelect` higher order component.
 * This subscribes the ModelSelect component to the state maintained via the
 * eventespresso/lists store.
 */
export default withSelect((select, ownProps) => {
	const { getQueryString, modelName, selectConfiguration } = ownProps;
	const queryString = getQueryString(ownProps.queryData);
	const { getEntities, isRequestingEntities } = select('eventespresso/lists');
	return {
		...ModelSelect.defaultProps,
		...ownProps,
		modelEntities: getEntities(modelName, queryString),
		selectConfiguration: {
			...ModelSelect.defaultProps.selectConfiguration,
			...selectConfiguration,
			isLoading: isRequestingEntities(modelName, queryString),
		},
	};
})(ModelSelect);
