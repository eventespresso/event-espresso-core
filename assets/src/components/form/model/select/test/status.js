import StatusSelect, { EditorStatusSelect } from '../status';
import { shallow, render } from 'enzyme';
import { QUERY_ORDER_ASC } from 'assets/src/data/model/base';

jest.mock('../../base/model-select', () => () => <span>StatusSelect</span>);

describe('StatusSelect()', () => {
	const expectedDefaultQueryData = {
		limit: 25,
		orderBy: 'statusCode',
		order: QUERY_ORDER_ASC,
	};
	it('matches snapshot with default props plus provided statusType', () => {
		const wrapper = shallow(<StatusSelect />);
		expect(wrapper).toMatchSnapshot();
	});
	it('matches expected default props plus provided statusType', () => {
		const wrapper = shallow(<StatusSelect />);
		expect(wrapper.prop('queryData')).toEqual(expectedDefaultQueryData);
		expect(wrapper.prop('selectConfiguration').placeholder).toEqual(
			'Select Statusâ€¦'
		);
		expect(wrapper.prop('modelName')).toEqual('status');
		expect(wrapper.prop('label')).toEqual('Select Status');
	});
});

describe('EditorStatusSelect()', () => {
	it('matches snapshot with default props', () => {
		const wrapper = render(<EditorStatusSelect />);
		expect(wrapper).toMatchSnapshot();
	});
});
