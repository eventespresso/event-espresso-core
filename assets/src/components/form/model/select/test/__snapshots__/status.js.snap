// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorStatusSelect() matches snapshot with default props 1`] = `
<div
  class="components-base-control css-wdf2ti-Wrapper e1puf3u0"
>
  <div
    class="components-base-control__field css-11vcxb9-StyledField e1puf3u1"
  >
    <label
      class="components-base-control__label css-pezhm9-StyledLabel e1puf3u2"
      for="inspector-select-status-control-0"
    >
      Select Status
    </label>
    <span>
      StatusSelect
    </span>
  </div>
</div>
`;

exports[`StatusSelect() matches snapshot with default props plus provided statusType 1`] = `
<Component
  getQueryString={[Function]}
  label="Select Status"
  modelName="status"
  optionsEntityMap={
    Object {
      "default": Object {
        "label": [Function],
        "value": "STS_ID",
      },
    }
  }
  queryData={
    Object {
      "limit": 25,
      "order": "ASC",
      "orderBy": "statusCode",
    }
  }
  selectConfiguration={
    Object {
      "defaultValue": undefined,
      "loadingMessage": [Function],
      "noOptionsMessage": [Function],
      "onChange": undefined,
      "placeholder": "Select Statusâ€¦",
    }
  }
/>
`;
